src/aspire/source/simulation.py

```diff
@@ -468,3 +475,132 @@ def eval_coords(self, mean_vol, eig_vols, coords_est):
         corr = inner / (norm_true * norm_est)
 
         return {"err": err, "rel_err": rel_err, "corr": corr}
+
+    def estimate_snr(self, *args, **kwargs):
+        """
+        See ImageSource.estimate_snr() for documentation.
+        """
+        # For clean images return infinite SNR.
+        # Note, relationship with CTF and other sim corruptions still isn't clear to me...
+        if self.noise_adder is None:
+            return np.inf
+
+        # For SNR of Simulations, use the theoretical noise variance
+        # known from the noise_adder instead of deriving from PSD.
+        return super().estimate_snr(noise_power=self.noise_adder.noise_var)
+
+    @classmethod
+    def from_snr(
```

Comment:
I view this just as `np.*_like` or `np.from_*`. Typically we would use a staticmethod (which we could do), but using a classmethod just makes it so that if we even subclass simulation, the inheritance will work right. 

I'm against the callback.  It would not make sense in any other context, and would be complicated (also making a circular coupling, sim->noiseadder->sim) .  We can use the local images fed to the noise filter at application time to derive a var dynamically per image, but this also seems pretty complicated (lazy eval).  OTOH It would have the benefit that each image would be closer the to target snr as opposed to estimating  the aggregate set of sim images.

If you would like me to implement it the other way I can look into it, but it will drag things on...

Reply:
I was able to do this, but it required a little more complexity than I hoped, and trimming some of the optional optimization from the PR mentioned elsewhere in the review.

I think this solution I came up with is okay though.  In the future if we get a collection of NoiseAdders, with minor code the logic would be inheritable...

LMK, thanks!

Reply:
Makes a lot more sense to me now. Thanks!